test/replication-luatest/gh_6126_auto_rejoin_test.lua

```diff
@@ -0,0 +1,51 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+local fio = require('fio')
+
+local g = t.group('gh_6126')
+
+g.before_each(function(cg)
+    cg.cluster = cluster:new({})
+
+    local box_cfg = {
+        replication = server.build_instance_uri('master'),
+    }
+    cg.replica = cg.cluster:build_and_add_server(
+        {alias = 'replica', box_cfg = box_cfg})
+
+    cg.master = cg.cluster:build_and_add_server(
+        {alias = 'master', box_cfg = box_cfg})
+
+    cg.master:start()
+end)
+
+g.after_each(function(cg)
+    cg.cluster:drop()
+end)
+
+g.test_auto_rejoin = function(cg)
+    -- Test that joining replica tolerates ER_READONLY errors from master
+    cg.master:exec(function() box.cfg{read_only = true} end)
+    cg.replica:start({wait_for_readiness = false})
+    local logfile = fio.pathjoin(cg.replica.workdir, 'replica.log')
+    t.helpers.retrying({}, function()
+        t.assert(cg.replica:grep_log('ER_READONLY', nil,
+            {filename = logfile}), 'Can\'t modify data on a read-only'..
+            'instance - box.cfg.read_only is true')
+
+        t.assert(g.master:exec(function()
+            return box.space._cluster:count() == 1
+        end),  'No join while master is read-only')
+    end)
+
+    -- Test that replica joins after master became writeable
+    cg.master:exec(function() box.cfg{read_only = false} end)
+    cg.replica:wait_for_readiness()
+    t.helpers.retrying({}, function()
+        t.assert(g.master:exec(function()
+            return box.space._cluster:count() == 2
+        end), 'Join after master became writeable')
+        cg.replica:assert_follows_upstream(1)
+    end)
```

Comment:
Please, add more test cases. For example, what if the master is writeable, but the access is denied?
```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3313,
}
s1 = box.schema.create_space('test1', {engine = 'vinyl'})
_ = s1:create_index('pk')
s1:replace{1}
s2 = box.schema.create_space('test2')
_ = s2:create_index('pk')
s2:replace{2}
box.snapshot()
s1:replace{3}
s2:replace{4}

-- Step 3
--
box.schema.user.grant('guest', 'super')

--
-- Instance 2
--
-- Step 2
--
box.cfg{replication = 3313}
```
It works, but would be good to have it in the tests.
Also, please, add more data to the master instance - a vinyl space, a memtx space, make a snapshot, write more data, that would ensure vinyl has created some files and has in-memory data too.

Reply:
Nope, I don't see it on the branch.

Reply:
Done